### PR TITLE
Fix tracing source run

### PR DESCRIPTION
**Pull request by WeichenXu123**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/WeichenXu123/mlflow/pull/13610?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/13610/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 13610
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

Fix tracing source run:

In langchain model inference, it might create worker threads to invoke inference requests,
but currently the "active run" becomes thread-local (since commit https://github.com/mlflow/mlflow/commit/487db1f6c9b62c40ccfcba76c25defdabbcc4acc  )

but tracing routine need to attach source run in every inference worker thread, but `mlflow.active_run()` returns None because active run is created in another thread, this causes source run information lost.

To address it, I edited code in `mlflow/tracing/processor/mlflow.py`.

### How is this PR tested?

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [x] Manual tests

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

### CODE REVIEW

File: mlflow/tracing/processor/mlflow.py [hunk 1]
```diff
@@ -86,10 +86,26 @@ def on_start(self, span: OTelSpan, parent_context: Optional[Context] = None):
...
         # If the span is started within an active MLflow run, we should record it as a trace tag
-        if run := mlflow.active_run():
+        # Note `MLflow.active_run()` can only get thread-local active run,
+        # but tracing routine might be applied to model inference worker threads
+        # (e.g. langchain model batch inference),
+        # and we only support global mode tracing,
+        # so the following code checks all active runs in all threads to get the latest active run
+        # as the tracing source run.
+        latest_active_run = None
+        for active_run_stack in _active_run_stack._value_dict.values():
```
Comment (B-Step62) [inline mlflow/tracing/processor/mlflow.py:101]: Is there any chance `_active_run_stack` will also be updated to thread local? It seems the stack is used for determining parent run in nested case, then it might be a scope of thread-safety?
https://github.com/WeichenXu123/mlflow/blob/5da87886ab760e61b5add6e175b4490397378c80/mlflow/tracking/fluent.py#L396-L398

File: mlflow/langchain/api_request_parallel_processor.py [hunk 2]
```diff
@@ -267,23 +266,11 @@ def process_api_requests(
...
-                    status_tracker=status_tracker, callback_handlers=callback_handlers
-                )
-
             # if enough capacity available, call API
             if next_request:
                 # call API
                 executor.submit(
-                    call_api,
-                    requester=next_request,
+                    next_request.call_api,
```
Comment (B-Step62) [inline mlflow/langchain/api_request_parallel_processor.py:273]: what is this change for?

File: mlflow/tracing/processor/mlflow.py [hunk 3]
```diff
@@ -86,10 +86,26 @@ def on_start(self, span: OTelSpan, parent_context: Optional[Context] = None):
...
     def _start_trace(self, span: OTelSpan, start_time_ns: Optional[int]) -> TraceInfo:
+        from mlflow.tracking.fluent import _active_run_stack
+
         experiment_id = get_otel_attribute(span, SpanAttributeKey.EXPERIMENT_ID)
         metadata = {TRACE_SCHEMA_VERSION_KEY: str(TRACE_SCHEMA_VERSION)}
         # If the span is started within an active MLflow run, we should record it as a trace tag
-        if run := mlflow.active_run():
+        # Note `MLflow.active_run()` can only get thread-local active run,
+        # but tracing routine might be applied to model inference worker threads
+        # (e.g. langchain model batch inference),
```
Comment (B-Step62) [inline mlflow/tracing/processor/mlflow.py:96]: Does "batch inference" here mean `chain.batch()` in LangChain or our `predict` method (api_request_parallel_processor)?

File: mlflow/utils/thread_utils.py [hunk 4]
```diff
@@ -42,3 +44,4 @@ def set(self, value):
         Set a value for the thread-local variable.
         """
         self.thread_local.value = (value, os.getpid())
+        self._value_dict[threading.currentThread().ident] = value
```
Comment (B-Step62) [inline mlflow/utils/thread_utils.py:47]: Is this change related to this PR?

File: mlflow/tracing/processor/mlflow.py (hunk 1 above)
Comment (WeichenXu123) [reply mlflow/tracing/processor/mlflow.py:101]: before this commit https://github.com/mlflow/mlflow/commit/487db1f6c9b62c40ccfcba76c25defdabbcc4acc

all threads shares the same `_active_run_stack`, i.e., even it is a stack, all threads can push runs into it, it is buggy in multiple thread case.

File: mlflow/langchain/api_request_parallel_processor.py (hunk 2 above)
Comment (WeichenXu123) [reply mlflow/langchain/api_request_parallel_processor.py:273]: This part change is introduced in this commit https://github.com/mlflow/mlflow/commit/487db1f6c9b62c40ccfcba76c25defdabbcc4acc

I revert it in this PR, because this part change can't fully address tracing source run lost issue (it only prevents tracing source run lost for MLflow pyfunc model)

File: mlflow/tracing/processor/mlflow.py (hunk 3 above)
Comment (WeichenXu123) [reply mlflow/tracing/processor/mlflow.py:96]: it means both of the cases you listed above

File: mlflow/utils/thread_utils.py (hunk 4 above)
Comment (WeichenXu123) [reply mlflow/utils/thread_utils.py:47]: Yes. `. _value_dict` attribute is used in `_start_trace` in this PR.

File: mlflow/tracing/processor/mlflow.py (hunk 3 above)
Comment (B-Step62) [reply mlflow/tracing/processor/mlflow.py:96]: Could you list two explicitly?

File: mlflow/langchain/api_request_parallel_processor.py (hunk 2 above)
Comment (B-Step62) [reply mlflow/langchain/api_request_parallel_processor.py:273]: ah I see, thanks for the explanation!

File: mlflow/utils/thread_utils.py [hunk 5]
```diff
@@ -45,4 +45,10 @@ def set(self, value):
         Set a value for the thread-local variable.
         """
         self.thread_local.value = (value, os.getpid())
-        self._value_dict[threading.currentThread().ident] = value
+        self.__global_thread_values[threading.currentThread().ident] = value
+
+    def get_global(self):
```
Comment (B-Step62) [inline mlflow/utils/thread_utils.py:50]:
```suggestion
    def get_globals(self) -> Dict[str, Any]:
```

File: mlflow/tracking/fluent.py [hunk 6]
```diff
@@ -613,6 +613,15 @@ def last_active_run() -> Optional[Run]:
     return get_run(last_active_run_id)
 
 
+def _get_latest_active_run():
+    latest_active_run = None
+    for active_run_stack in _active_run_stack._value_dict.values():
```
Comment (B-Step62) [inline mlflow/tracking/fluent.py:618]:
```
for active_run_stack in _active_run_stack.get_globals().values():
```

File: mlflow/tracking/fluent.py [hunk 7]
```diff
@@ -613,6 +613,15 @@ def last_active_run() -> Optional[Run]:
     return get_run(last_active_run_id)
 
 
+def _get_latest_active_run():
```
Comment (B-Step62) [inline mlflow/tracking/fluent.py:616]: Can you update the name to indicate this is global scope, and add docstring that describes the difference from `last_active_run()`?

Comment (B-Step62) [reply mlflow/tracking/fluent.py:616]: Can we also update docstring of `last_active_run` to state it is thread-local, because it's user-facing API.

File: tests/langchain/test_langchain_autolog.py [hunk 8]
```diff
@@ -771,6 +771,19 @@ def test_langchain_autolog_callback_injection_in_batch(invoke_arg, config, async
...
+
+    model = create_openai_llmchain()
+    input = {"product": "MLflow"}
+    with mlflow.start_run() as run:
+        model.batch([input] * 2)
+
+    traces = get_traces()
+    for trace in traces:
+        assert trace.info.request_metadata[TraceMetadataKey.SOURCE_RUN] == run.info.run_id
+
```
Comment (B-Step62) [inline tests/langchain/test_langchain_autolog.py:785]:
```suggestion
def test_tracing_source_run_in_batch():
    mlflow.langchain.autolog(log_models=True)

    model = create_openai_llmchain()
    input = {"product": "MLflow"}
    with mlflow.start_run() as run:
        model.batch([input] * 2)

    traces = get_traces()
    for trace in traces:
        assert trace.info.request_metadata[TraceMetadataKey.SOURCE_RUN] == run.info.run_id

    pyfunc_model = mlflow.pyfunc.load_model(run.info.run_id)
    with mlflow.start_run() as run_2:
        pyfunc_model.predict("input")
        
    trace = mlflow.get_last_active_trace()
    assert trace.info.request_metadata[TraceMetadataKey.SOURCE_RUN] == run_2.info.run_id
```

Let's also test the pyfunc prediction.

File: mlflow/tracking/fluent.py [hunk 9]
```diff
@@ -526,6 +526,8 @@ def active_run() -> Optional[ActiveRun]:
     """
     Get the currently active ``Run``, or None if no such run exists.
 
+    **Note**: The API is thread-local.
```
Comment (B-Step62) [inline mlflow/tracking/fluent.py:529]:
```suggestion
    .. attention::

        This API is **thread-local** and returns only the active run in the current thread. If your application is multi-threaded and a run is started in a different thread, this API will not retrieve that run.
```

"thread-local" alone is not clear enough for users especially data scientists

Comment (B-Step62) [reply mlflow/tracking/fluent.py]: Can we add this to `last_active_run` too

File: mlflow/tracking/fluent.py (hunk 7 above)
Comment (B-Step62) [reply mlflow/tracking/fluent.py:621]: I don't think `_get_latest_active_run` indicates global vs thread-local. What about `_get_global_active_run`?

File: tests/langchain/test_langchain_autolog.py [hunk 10]
```diff
@@ -771,6 +771,26 @@ def test_langchain_autolog_callback_injection_in_batch(invoke_arg, config, async
             assert set(handler.logs) == {"chain_start", "chain_end"}
 
 
+def test_tracing_source_run_in_batch():
+    mlflow.langchain.autolog()
```
Comment (B-Step62) [inline tests/langchain/test_langchain_autolog.py:775]:
```suggestion
    mlflow.langchain.autolog(log_models=True)
```

You need to pass `log_models=True` otherwise a model won't be logged from autolog.

File: mlflow/tracking/fluent.py [hunk 11]
```diff
@@ -613,6 +615,18 @@ def last_active_run() -> Optional[Run]:
     return get_run(last_active_run_id)
 
 
+def _get_latest_active_run():
+    """
+    Check active run stack in all threads, find the latest active run and return it.
```
Comment (B-Step62) [inline mlflow/tracking/fluent.py:623]:
```suggestion
    Get active run from global context by checking all threads. The `mlflow.active_run` API only returns 
    active run from current thread. This API is useful for the case where one needs to get a run started
    from a separate thread.
```

File: mlflow/tracing/processor/mlflow.py [hunk 12]
```diff
@@ -86,10 +85,21 @@ def on_start(self, span: OTelSpan, parent_context: Optional[Context] = None):
...
         # If the span is started within an active MLflow run, we should record it as a trace tag
-        if run := mlflow.active_run():
+        # Note `MLflow.active_run()` can only get thread-local active run,
+        # but tracing routine might be applied to model inference worker threads
+        # in the following cases:
+        #  - langchain model `chain.batch` which uses thread pool to spawn workers.
+        #  - MLflow langchain pyfunc model `predict` which calls `api_request_parallel_processor`.
+        # and we only support global mode tracing,
+        # so the following code checks all active runs in all threads to get the latest active run
+        # as the tracing source run.
```
Comment (B-Step62) [inline mlflow/tracing/processor/mlflow.py:100]:
```suggestion
        # Therefore, we use `_get_global_active_run()` instead to get the active run from
        # all threads and set it as the tracing source run.
```

**Review by B-Step62 — APPROVED** [review]
LGTM once the remaining comments are addressed and x-version tests pass🙂

File: mlflow/utils/thread_utils.py [hunk 13]
```diff
@@ -45,4 +46,10 @@ def set(self, value):
         Set a value for the thread-local variable.
         """
         self.thread_local.value = (value, os.getpid())
-        self._value_dict[threading.currentThread().ident] = value
+        self.__global_thread_values[threading.currentThread().ident] = value
```
Comment (harupy) [inline mlflow/utils/thread_utils.py:49]:
```suggestion
        self._global_thread_values[threading.currentThread().ident] = value
```

you don't need two underscores

Comment (WeichenXu123) [reply mlflow/utils/thread_utils.py:49]: oh 2 underscores is suggested by @B-Step62 (for true private) , I am ok for either 1 or 2

Comment (B-Step62) [reply mlflow/utils/thread_utils.py:49]: I'm actually inclined to two underscores to protect this more strongly - writing to this directly or mutating it can be a source of confusing bug. WDYT?

File: mlflow/tracing/processor/mlflow.py [hunk 14]
```diff
@@ -86,10 +85,20 @@ def on_start(self, span: OTelSpan, parent_context: Optional[Context] = None):
         span.set_attribute(SpanAttributeKey.REQUEST_ID, json.dumps(request_id))
 
     def _start_trace(self, span: OTelSpan, start_time_ns: Optional[int]) -> TraceInfo:
+        from mlflow.tracking.fluent import _get_latest_active_run
+
         experiment_id = get_otel_attribute(span, SpanAttributeKey.EXPERIMENT_ID)
         metadata = {TRACE_SCHEMA_VERSION_KEY: str(TRACE_SCHEMA_VERSION)}
         # If the span is started within an active MLflow run, we should record it as a trace tag
-        if run := mlflow.active_run():
+        # Note `MLflow.active_run()` can only get thread-local active run,
```
Comment (harupy) [inline mlflow/tracing/processor/mlflow.py:93]:
```suggestion
        # Note `mlflow.active_run()` can only get thread-local active run,
```

File: mlflow/utils/thread_utils.py (hunk 13 above)
Comment (harupy) [reply mlflow/utils/thread_utils.py:49]: makes sense, let's keep two underscore here.

File: mlflow/utils/thread_utils.py [hunk 15]
```diff
@@ -45,4 +46,10 @@ def set(self, value):
...
+
+    def get_global(self) -> Dict[str, Any]:
```
Comment (harupy) [inline mlflow/utils/thread_utils.py:51]:
```suggestion
    def get_all_thread_values(self) -> Dict[str, Any]:
```

for clarity.

File: mlflow/tracking/fluent.py [hunk 16]
```diff
@@ -613,6 +618,20 @@ def last_active_run() -> Optional[Run]:
...
+    Get active run from global context by checking all threads. The `mlflow.active_run` API
+    only returns active run from current thread. This API is useful for the case where one
+    needs to get a run started from a separate thread.
+    """
+    latest_active_run = None
+    for active_run_stack in _active_run_stack.get_global().values():
+        for active_run in active_run_stack:
+            if latest_active_run is None or active_run.start_time > latest_active_run.start_time:
+                latest_active_run = active_run
+    return latest_active_run
```
Comment (harupy) [inline mlflow/tracking/fluent.py:632]: Can we simplify this function using max?

Comment (harupy) [reply mlflow/tracking/fluent.py:632]: /ai suggest the code.

Comment (harupy) [reply mlflow/tracking/fluent.py:632]: @WeichenXu123 Not a blocker.

File: mlflow/tracing/processor/mlflow.py [hunk 17]
```diff
@@ -86,10 +85,20 @@ def on_start(self, span: OTelSpan, parent_context: Optional[Context] = None):
...
-        if run := mlflow.active_run():
+        # Note `mlflow.active_run()` can only get thread-local active run,
+        # but tracing routine might be applied to model inference worker threads
+        # in the following cases:
+        #  - langchain model `chain.batch` which uses thread pool to spawn workers.
+        #  - MLflow langchain pyfunc model `predict` which calls `api_request_parallel_processor`.
+        # Therefore, we use `_get_global_active_run()` instead to get the active run from
+        # all threads and set it as the tracing source run.
+        latest_active_run = _get_latest_active_run()
+        if run := latest_active_run:
```
Comment (harupy) [inline mlflow/tracing/processor/mlflow.py:101]:
```suggestion
        if run := _get_latest_active_run():
```


Is `latest_active_run` used after this line?

File: mlflow/utils/thread_utils.py (hunk 13 above)
Comment (harupy) [inline mlflow/utils/thread_utils.py:49]:
```suggestion
        self.__global_thread_values[threading.currentThread().get_ident()] = value
```

The official python doc doesn't mention an `ident` attribute. 

https://docs.python.org/3/library/threading.html

Let's use something that's mentioned for safety.

**Review by harupy — APPROVED** [review]
Left some comments, otherwise LGTM!

File: mlflow/tracing/processor/mlflow.py [hunk 18]
```diff
@@ -86,10 +85,20 @@ def on_start(self, span: OTelSpan, parent_context: Optional[Context] = None):
...
-        if run := mlflow.active_run():
+        # Note `mlflow.active_run()` can only get thread-local active run,
+        # but tracing routine might be applied to model inference worker threads
+        # in the following cases:
+        #  - langchain model `chain.batch` which uses thread pool to spawn workers.
+        #  - MLflow langchain pyfunc model `predict` which calls `api_request_parallel_processor`.
+        # Therefore, we use `_get_global_active_run()` instead to get the active run from
+        # all threads and set it as the tracing source run.
+        run = _get_latest_active_run()
+        if run:
```
Comment (harupy) [inline mlflow/tracing/processor/mlflow.py:101]:
```suggestion
        if run := _get_latest_active_run():
```

File: mlflow/tracking/fluent.py [hunk 19]
```diff
@@ -613,6 +618,18 @@ def last_active_run() -> Optional[Run]:
...
+def _get_latest_active_run():
+    """
+    Get active run from global context by checking all threads. The `mlflow.active_run` API
+    only returns active run from current thread. This API is useful for the case where one
+    needs to get a run started from a separate thread.
+    """
+    all_active_runs = [
+        run for run_stack in _active_run_stack.get_all_thread_values().values() for run in run_stack
+    ]
+    return max(all_active_runs, key=lambda run: run.start_time)
```
Comment (harupy) [inline mlflow/tracking/fluent.py:630]: max throws if `all_active_runs` is empty.
